web/packages/teleport/src/JoinTokens/JoinTokens.tsx

```diff
@@ -0,0 +1,399 @@
+
+import { useTeleport } from 'teleport';
+import useResources from 'teleport/components/useResources';
+
+import {
+  FeatureBox,
+  FeatureHeader,
+  FeatureHeaderTitle,
+} from 'teleport/components/Layout';
+import { JoinToken } from 'teleport/services/joinToken';
+import { Resource, KindJoinToken } from 'teleport/services/resources';
+import ResourceEditor from 'teleport/components/ResourceEditor';
+
+function makeTokenResource(token: JoinToken): Resource<KindJoinToken> {
+  return {
+    id: token.id,
+    name: token.safeName,
+    kind: 'join_token',
+    content: token.content,
+  };
+}
+
+export const JoinTokens = () => {
+  const ctx = useTeleport();
+  const [tokenToDelete, setTokenToDelete] = useState<JoinToken | null>(null);
+  const [joinTokensAttempt, runJoinTokensAttempt, setJoinTokensAttempt] =
+    useAsync(async () => await ctx.joinTokenService.fetchJoinTokens());
+
+  const resources = useResources(
+    joinTokensAttempt.data?.items.map(makeTokenResource) || [],
+    { join_token: '' } // we are only editing for now, so template can be empty
+  );
+
+  async function handleSave(content: string): Promise<void> {
+    const token = await ctx.joinTokenService.upsertJoinToken({ content });
+    let items = [...joinTokensAttempt.data.items];
+    if (resources.status === 'creating') {
+      items.push(token);
+    } else {
+      let tokenExistsInPreviousList = false;
+      const newItems = items.map(item => {
+        if (item.id === token.id) {
+          tokenExistsInPreviousList = true;
+          return token;
+        }
+        return item;
+      });
+      // in the edge case that someone only edits the name of the token, it will return
+      // a "new" token via the upsert, and therefore should be treated as a new token
+      if (!tokenExistsInPreviousList) {
+        newItems.push(token);
+      }
+      items = newItems;
+    }
+    setJoinTokensAttempt({
+      data: { ...joinTokensAttempt.data, items },
+      status: 'success',
+      statusText: '',
+    });
+  }
+
+  const [deleteTokenAttempt, runDeleteTokenAttempt] = useAsync(
+    async (token: string) => {
+      await ctx.joinTokenService.deleteJoinToken(token);
+      setJoinTokensAttempt({
+        status: 'success',
+        statusText: '',
+        data: {
+          items: joinTokensAttempt.data.items.filter(t => t.id !== token),
+        },
+      });
+      setTokenToDelete(null);
+    }
+  );
+
+  useEffect(() => {
+    runJoinTokensAttempt();
+  }, []);
+
+  return (
+    <FeatureBox>
+      <FeatureHeader
+        css={`
+          // TODO (avatus) remove all border-bottom from FeatureHeader, requested by design
+          border-bottom: none;
+        `}
+        alignItems="center"
+      >
+        <FeatureHeaderTitle>Join Tokens</FeatureHeaderTitle>
+      </FeatureHeader>
+      <Box>
+        {deleteTokenAttempt.status === 'error' && (
+          <Alert kind="danger">{deleteTokenAttempt.statusText}</Alert>
+        )}
```

Comment:
what about the other failure modes? like fetching and upserting

Reply:
Thnks, i forgot to add alerts for those!

Reply:
I added the alert for fetching, but the error for upserting is shown in the yaml editor. i will update this when we add the forms